### PR TITLE
Various fixes and improvements

### PR DESCRIPTION
**Pull request by daandemeyer**

### CODE REVIEW

File: src/shared/udev-util.c [hunk 1]
```diff
@@ -165,9 +165,11 @@ static int device_monitor_handler(sd_device_monitor *monitor, sd_device *device,
         if (data->devlink) {
                 const char *devlink;
 
-                FOREACH_DEVICE_DEVLINK(device, devlink)
+                FOREACH_DEVICE_DEVLINK(device, link) {
+                        devlink = link;
```
Comment (yuwata) [inline src/shared/udev-util.c]: The assignment is not necessary.

File: src/udev/udevadm-info.c [hunk 2]
```diff
@@ -250,7 +252,7 @@ static int print_record(sd_device *device, const char *prefix) {
         if (sd_device_get_driver(device, &str) >= 0)
                 printf("%sV: %s%s%s\n", prefix, ansi_highlight_yellow4(), str, ansi_normal());
 
-        FOREACH_DEVICE_PROPERTY(device, str, val)
+        FOREACH_DEVICE_PROPERTY(device, key, val)
                 printf("%sE: %s=%s\n", prefix, str, val);
```
Comment (yuwata) [inline src/udev/udevadm-info.c]: `str` -> `key`

**Review by yuwata — CHANGES_REQUESTED** [review]

**Comment by DaanDeMeyer** [comment]
Added one more commit to make `json_variant_unref_many()` usable with `CLEANUP_ARRAY()` and another commit to do a simple rename, PTAL!

File: src/basic/terminal-util.h [hunk 3]
```diff
@@ -49,6 +49,7 @@
 
 /* Underlined */
 #define ANSI_GREY_UNDERLINE              "\x1B[0;4;38;5;245m"
+#define ANSI_BRIGHT_BLACK_UNDERLINE      "\x1B[0;4;90m"
```
Comment (poettering) [inline src/basic/terminal-util.h:52]: what is "bright black"???

Comment (daandemeyer) [reply src/basic/terminal-util.h:52]: No clue but it's a color we have defined so it needs a specific underline version as well

Comment (bluca) [reply src/basic/terminal-util.h:52]: it's the opposite of 'dark white', duh

File: src/shared/logs-show.c [hunk 4]
```diff
@@ -942,6 +942,19 @@ struct json_data {
         JsonVariant* values[];
 };
 
+static struct json_data* json_data_free(struct json_data *d) {
+        json_variant_unref(d->name);
```
Comment (bluca) [inline src/shared/logs-show.c:949]: assert(d)

Comment (bluca) [reply src/shared/logs-show.c:949]: or maybe better return NULL if d is NULL

**Review by bluca — APPROVED** [review]

File: src/shared/logs-show.c (hunk 4 above)
Comment (daandemeyer) [reply src/shared/logs-show.c:949]: Ouch

File: src/shared/logs-show.c [hunk 5]
```diff
@@ -942,6 +942,22 @@ struct json_data {
         JsonVariant* values[];
 };
 
+static struct json_data* json_data_free(struct json_data *d) {
+        if (!d)
+                return NULL;
+
+        json_variant_unref(d->name);
+
+        for (size_t i = 0; i < d->n_values; i++)
```
Comment (YHNdnzj) [inline src/shared/logs-show.c]: You can use `FOREACH_ARRAY` for this.

File: src/shared/tpm2-util.c [hunk 6]
```diff
@@ -3862,7 +3862,9 @@ int tpm2_make_pcr_json_array(uint32_t pcr_mask, JsonVariant **ret) {
         r = 0;
 
 finish:
-        json_variant_unref_many(pcr_array, n_pcrs);
+        for (size_t i = 0; i < n_pcrs; i++)
```
Comment (YHNdnzj) [inline src/shared/tpm2-util.c]: Same here.

File: src/udev/udevadm-query.c [hunk 7]
```diff
@@ -0,0 +1,148 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later */
```
Comment (YHNdnzj) [inline src/udev/udevadm-query.c]: Doesn't belong to this PR. Rebase failure?

File: src/udev/udevadm-query.c [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+#include "format-table.h"
+#include "terminal-util.h"
+
+static JsonFormatFlags arg_json_format_flags = JSON_FORMAT_OFF;
+static PagerFlags arg_pager_flags = 0;
+static bool arg_legend = true;
+
+static int help(void) {
+        printf("%s wait [OPTIONS] DEVICE [DEVICE…]\n\n"
+               "Wait for devices or device symlinks being created.\n\n"
```
Comment (YHNdnzj) [inline src/udev/udevadm-query.c]: (help message is not updated BTW)

**Review by YHNdnzj — CHANGES_REQUESTED** [review]

File: src/udev/udevadm-query.c (hunk 7 above)
Comment (daandemeyer) [reply src/udev/udevadm-query.c]: Consider it a sneak peek ;)

**Review by YHNdnzj — APPROVED** [review]

**Comment by DaanDeMeyer** [comment]
I forgot two iterators in device-util so I declared those inline as well

**Pull request merged**